rfcs/2022-05-17-11532-chronicle-sink.md

```diff
@@ -0,0 +1,237 @@
+ensure the event is well formed - either the source will receive the data in the
+correct structure, or it will be transformed into such before reaching the sink.
+
+#### Unstructured log data events
+
+The `gcp_chronicle_unstructured` sink.
+
+Unstructured log events are events in plain text with no strict schema attached.
+
+```json
+{
+  "log_text": "26-Feb-2019 13:35:02.187 client 10.120.20.32#4238: query: altostrat.com IN A + (203.0.113.102)",
+  "ts_epoch_microseconds": 1551188102187000
+}
+```
+
+Events are batched and sent in a JSON object that includes the `customer_id` and
+a `log_type`:
+
+```json
+{
+  "customer_id": "c8c65bfa-5f2c-42d4-9189-64bb7b939f2c",
+  "log_type": "BIND_DNS",
+  "entries": [...]
+}
+```
+
+`log_type` has to be one of the values returned by the [logtypes endpoint](https://cloud.google.com/chronicle/docs/reference/ingestion-api#logtypes).
+
+
+If `event_type == "unstructured"` we will require another configuration field -
+`log_type` to be set. This field will be templatable. The field can be removed
+from the final event via `encoding.exclude_fields`.
+
+The encoding of the message (text or json) can be set via an `encoding` field.
+
+##### Schema
+
+A UDM message has the following sections. (To keep this RFC succinct, this is
+not a comprehensive list, the full list can be found
+[here](https://cloud.google.com/chronicle/docs/reference/udm-field-list#udm_event_data_model).)
+
+- *Metadata*
+  Contains metadata about the event. The fields available for metadata can be
+  found [here](https://cloud.google.com/chronicle/docs/reference/udm-field-list#metadata).
+  Mandatory fields are `event_type` and `event_timestamp`.
+
+  The value of `event_type` has repercussions for which fields are mandatory
+  in the rest of the event. See [here](https://cloud.google.com/chronicle/docs/unified-data-model/udm-usage#required_and_optional_fields_based_on_event_type)
+  for the list. For example, if `event_type == "FILE_COPY"` then `src.file`
+  becomes mandatory. **Currently it is not possible to validate a field based
+  on the value of another field within Vector schema, so this will need to
+  be left for future improvement.
+
+- *Principal*
+  The principal is the entity that originates the event. This field is mandatory
+  and can contain any of the fields defined
+  [here](https://cloud.google.com/chronicle/docs/reference/udm-field-list#noun).
+
+- *Src*
+  The source entity being acted upon. Depending on the `event_type` this field
+  is optional. Can contain any of the fields defined
+  [here](https://cloud.google.com/chronicle/docs/reference/udm-field-list#noun).
+
+- *Target*
+  The target entity being acted upon. Depending on the `event_type` this field
+  is optional. Can contain any of the fields defined
+  [here](https://cloud.google.com/chronicle/docs/reference/udm-field-list#noun).
+
+### Implementation
+
+Both event types require different batching strategies.
```

Comment:
@tobz I thought that too, but I'm not sure avoiding unstructured payloads is a good idea. Chronicle has a lot of first class support for unstructured log events, you can see how that [here](https://cloud.google.com/chronicle/docs/data-ingestion-flow). I'm not sure what the right answer is, but I have to assume some users will want to send unstructured data due to the all of automatic detection and parsing Chronicle provides.

Reply:
I'm suggesting that we can look at any given event, and say:
- does it match the UDM schema? batch/send it as a UDM payload
- does it have the `log_type` so we can treat it as unstructured? batch/send it as an unstructured payload
- does it have neither? emit a warning/error

The code to perform the above logic will have to exist, whether it's in a single sink, or split across two sinks.

Reply:
I think a user should really know up front whether they are intending to send UDM or unstructured. The `log_type` may not be a part of the message, the user may want to just configure all events to go under a given log type. If we implemented it this way we risk sending an malformed UDM event as unstructured rather than erroring. 

I would be prefer to let the user explicitly state what they are sending and us making sure it conforms to that. Especially since this is security events that are being sent, so the cost of it going wrong is much higher..

Reply:
I agree here for the initial implementation: make the user specify whether data should be sent to the unstructured endpoint, or the UDM endpoint. Maybe we can do work in the future to performantly figure that out on the fly, but until then, yeah, we should strive to not surprise the user with how we're routing their security events.